phabricatoremails/render.py

```diff
@@ -0,0 +1,420 @@
+                        [
+                            parse_inline_comment(inline)
+                            for inline in body["inlineComments"]
+                        ],
+                        body["transactionLink"],
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                    )
+                batch.target_all(body.reviewers, "abandoned")
+            elif kind == "revision-reclaimed":
+                if is_secure:
+                    body = SecureRevisionReclaimed(
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                        body["commentCount"],
+                        body["transactionLink"],
+                    )
+                else:
+                    body = RevisionReclaimed(
+                        body.get("mainComment"),
+                        [
+                            parse_inline_comment(inline)
+                            for inline in body["inlineComments"]
+                        ],
+                        body["transactionLink"],
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                    )
+                batch.target_all(body.reviewers, "reclaimed")
+            elif kind == "revision-created":
+                if is_secure:
+                    body = SecureRevisionCreated(
+                        [parse_reviewer(reviewer) for reviewer in body["reviewers"]],
+                    )
+                else:
+                    body = RevisionCreated(
+                        [parse_affected_file(file) for file in body["affectedFiles"]],
+                        [parse_reviewer(reviewer) for reviewer in body["reviewers"]],
+                    )
+                for reviewer in body.reviewers:
+                    batch.target(reviewer.recipient, "created", reviewer=reviewer)
+            else:
+                raise ParseError("unexpected revision event kind: {}".format(kind))
+
+            actor_name = event["actorName"]
+
+            def query_thread(revision_id):
+                existing = (
+                    db.query(Thread).filter_by(phabricator_id=revision_id).first()
+                )
+                if existing:
+                    return existing
+
+                new = Thread(phabricator_id=revision_id, thread_size=0)
+                db.add(new)
+                return new
+
+            if is_secure:
+                revision = parse_secure_revision(event["revision"])
+                thread = query_thread(revision.id)
+                outgoing_emails.extend(
+                    batch.process_secure(
+                        template_store,
+                        revision,
+                        actor_name,
+                        thread.thread_size,
+                        timestamp,
+                        body,
+                    )
+                )
+                thread.thread_size += 1
```

Comment:
I don't think we want to create a new session every time we want to create/modify a `Thread`.
I placed the session management in `Progress` because we want the entire batch of events to be handled in a single transaction.

This simplifies our error handling. Let's imagine that, for some reason, writing to the database fails, This _should not happen_, and is a failure case. We report the error to Sentry and stop the application, because there's an issue that needs to be solved by a human.

If, when this error happens:
1. We roll back the changes for the batch of events, that means that when the service runs again, we can continue where we left off - re-fetch the entire batch of events from Phabricator, send the emails, continue to the next batch, and so on
2. We roll back the changes just for the current event, then we need to manually make sure that when we start the service again, we don't re-send the first emails of the batch that sent successfully. This requires some additional database work to track our state, which may be impossible to do since we're in a failure state where DB requests were failing.